src/batch/src/executor/join/nested_loop_join.rs

```diff
@@ -213,387 +197,382 @@ impl BoxedExecutorBuilder for NestedLoopJoinExecutor {
+                        if let Some(output_chunk) = output {
+                            yield output_chunk
+                        }
+                    }
+                }
             }
-        }
-        // Append (probed_row, None) to chunk builder if current row finished probing and do not
-        // find any match.
-        if ret.cur_row_finished && !self.probe_side_source.get_cur_row_matched() {
-            assert!(ret.chunk.is_none());
-            let mut probe_datum_refs = self
-                .probe_side_source
-                .get_current_row_ref()
-                .unwrap()
-                .values()
-                .collect_vec();
-            for _ in 0..self.build_table.get_schema().fields.len() {
-                probe_datum_refs.push(None);
+            if !matched {
+                let datum_refs = left_row
+                    .values()
+                    .chain(repeat_n(None, data_types.len() - left_data_types.len()));
+                if let Some(chunk) = chunk_builder.append_one_row_from_datum_refs(datum_refs)? {
+                    yield chunk
+                }
             }
-            let one_row_chunk =
-                self.convert_datum_refs_to_chunk(&probe_datum_refs, 1, &self.schema.data_types())?;
-            return Ok(ProbeResult {
-                cur_row_finished: true,
-                chunk: Some(one_row_chunk),
-            });
         }
-        Ok(ret)
+        if let Some(chunk) = chunk_builder.consume_all()? {
+            yield chunk
+        }
     }
 
-    fn do_left_semi_join(&mut self) -> Result<ProbeResult> {
-        let mut ret = self.do_inner_join()?;
-        if let Some(inner) = ret.chunk.as_ref() {
-            if inner.cardinality() > 0 {
-                self.probe_side_source.set_cur_row_matched(true);
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_left_semi_join(
+        data_types: Vec<DataType>,
+        join_expr: BoxedExpression,
+        left: Vec<DataChunk>,
+        right: Vec<DataChunk>,
+    ) {
+        let mut chunk_builder = DataChunkBuilder::with_default_size(data_types.clone());
+        for left_row in left.iter().flat_map(|chunk| chunk.rows()) {
+            let mut matched = false;
+            for right_chunk in right.iter() {
+                let left_chunk =
+                    Self::convert_row_to_chunk(&left_row, right_chunk.capacity(), &data_types)?;
+                let mut chunk = Self::concatenate(&left_chunk, right_chunk);
+                let vis = join_expr.eval(&chunk)?.as_bool().iter().collect();
+                chunk.set_visibility(vis);
+                if chunk.cardinality() > 0 {
+                    matched = true;
+                    break;
+                }
+            }
+            if matched {
+                if let Some(chunk) = chunk_builder.append_one_row_ref(left_row)? {
+                    yield chunk
+                }
             }
         }
```

Comment:
This seem incorrect, output of semi join should be determined after all right side data has been processed.